opal/mca/pmix/pmix3x/pmix3x_component.c

```diff
@@ -97,13 +98,15 @@ static int external_register(void)
 
     asprintf(&pmix_library_version,
              "PMIx library version %s (embedded in Open MPI)", PMIx_Get_version());
+    tmp = pmix_library_version;
     (void) mca_base_component_var_register(component, "library_version",
                                            "Version of the underlying PMIx library",
                                            MCA_BASE_VAR_TYPE_STRING,
                                            NULL, 0, 0,
                                            OPAL_INFO_LVL_4,
                                            MCA_BASE_VAR_SCOPE_CONSTANT,
                                            &pmix_library_version);
+    free(tmp);
```

Comment:
This is correct. The MCA base duplicates the string because it becomes owned by the MCA base (to support MPI_T). If the original value was allocated it needs to be freed by the caller.

Reply:
Personal opinion: this is u.g.l.y ;) Why not simply taking ownership on the value and letting the upper level cope with that ?

Reply:
Because most string variables are constant and so I wanted the ability to set the initial value to a constant string.

Reply:
This is how I read the code :
 - `&pmix_library_version` is saved at line 1495 (note the `&`)
 - Then `pmix_library_version` is replaced with the strdup’ed string at line 1499.

Bottom line, the original `pmix_library_version` is no more referenced, that is why I saved it in `tmp` in order to `free(tmp)`

@hjelmn can you please advise ?

On second thought, a more elegant approach could be to have this handled in MCA via a new flag (no `strdup()` at all or `strdup(); free()`)

Reply:
@hjelmn but then why returning a strdup instead of letting the upper level to make a copy of its own string if/when needed. Instead, we can guarantee that the MCA will not modify and not free the string.

Reply:
Optimize for the common case and let the uncommon case have the extra code.

Reply:
This behavior is well-documented.

Reply:
@ggouaillardet is right, nobody's using the original pointer. Moreover, the storage object must be global so that the mbv_storage could point to something that remains valid during the entire execution.